src/Security/ZitadelAuthenticator.php

```diff
@@ -0,0 +1,125 @@
+use App\Service\IpResolver;
+use App\Service\UserManager;
+use App\Utils\Slugger;
+use Doctrine\ORM\EntityManagerInterface;
+use KnpU\OAuth2ClientBundle\Client\ClientRegistry;
+use KnpU\OAuth2ClientBundle\Security\Authenticator\OAuth2Authenticator;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Routing\RouterInterface;
+use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
+use Symfony\Component\Security\Core\Exception\AuthenticationException;
+use Symfony\Component\Security\Http\Authenticator\Passport\Badge\RememberMeBadge;
+use Symfony\Component\Security\Http\Authenticator\Passport\Badge\UserBadge;
+use Symfony\Component\Security\Http\Authenticator\Passport\Passport;
+use Symfony\Component\Security\Http\Authenticator\Passport\SelfValidatingPassport;
+
+class ZitadelAuthenticator extends OAuth2Authenticator
+{
+    public function __construct(
+        private readonly ClientRegistry $clientRegistry,
+        private readonly RouterInterface $router,
+        private readonly EntityManagerInterface $entityManager,
+        private readonly UserManager $userManager,
+        private readonly IpResolver $ipResolver,
+        private readonly Slugger $slugger,
+        private readonly UserRepository $userRepository,
+    ) {
+    }
+
+    public function supports(Request $request): ?bool
+    {
+        return 'oauth_zitadel_verify' === $request->attributes->get('_route');
+    }
+
+    public function authenticate(Request $request): Passport
+    {
+        $client = $this->clientRegistry->getClient('zitadel');
+        $slugger = $this->slugger;
+
+        $provider = $client->getOAuth2Provider();
+
+        $accessToken = $provider->getAccessToken('authorization_code', [
+            'code' => $request->query->get('code'),
+        ]);
+
+        $rememberBadge = new RememberMeBadge();
+        $rememberBadge = $rememberBadge->enable();
+
+        return new SelfValidatingPassport(
+            new UserBadge($accessToken->getToken(), function () use ($accessToken, $client, $slugger) {
+                /** @var ZitadelResourceOwner $zitadelUser */
+                $zitadelUser = $client->fetchUserFromToken($accessToken);
+
+                $existingUser = $this->entityManager->getRepository(User::class)->findOneBy(
+                    ['oauthZitadelId' => $zitadelUser->getId()]
+                );
+
+                if ($existingUser) {
+                    return $existingUser;
+                }
+
+                $user = $this->userRepository->findOneBy(['email' => $zitadelUser->getEmail()]);
+
+                if ($user) {
+                    $user->oauthZitadelId = $zitadelUser->getId();
+
+                    $this->entityManager->flush();
+
+                    return $user;
+                }
+
+                $username = $slugger->slug($zitadelUser->toArray()['username']);
```

Comment:
On my instance the `username` definitely exists, but `preferred_username` included the domain which isn't what I wanted (e.g. I'd get `thepaperpilot_at_domain.tld`)

Reply:
On [this page](https://zitadel.com/docs/apis/openidoauth/claims) in the docs it lists what fields are provided under which requests, and it has a row indicating it's provided when requested in the introspect request if the scope is specifically requested (the third column), which it is as part of the `profile` scope.
![image](https://github.com/MbinOrg/mbin/assets/3683148/ae0423ff-a47d-4645-9450-442f7dc524c3)

Reply:
hmm ok, it might as well be zitadel misconfiguration on my part. for the `username` containing the `-at-` part is probably because zitadel likes usernames in that format, but it looks like when creating a user via admin, you can just put in user name without `@domain.tld` part and it'll accept it just fine, and that's what I did when creating a test user for testing this patch, guess I'll just let you know that this can happen for possible investigation 

for the record, this is what I get from dumping the `$zitadelUser->toArray()` just before getting username

```php
$username = $slugger->slug(dump($zitadelUser->toArray())['username'])
```

```php
 array:9 [▼
  "sub" => "249646190008795139"
  "name" => "ZITADEL Admin"
  "given_name" => "ZITADEL"
  "family_name" => "Admin"
  "locale" => "en"
  "updated_at" => 1705296726
  "preferred_username" => "tama@zitadel.localhost"
  "email" => "tama@zitadel.localhost"
  "email_verified" => true
]
```

Reply:
I found out you were right and my code was actually preventing new users from authenticating